File: internal/repositories/codec/lzhuf.go
package codec

// LZHUF (method 0x80) — register-faithful port of NGI32.DLL sub_26F4E (0x80
// branch). LZSS + adaptive Huffman with NGI's own (non-canonical) position
// tables. Mirrors ../../../tools/lzhuf.py exactly. See ../../docs/02-compression.md.

const (
	lzN      = 4096
	lzF      = 60
	lzThresh = 2
	nChar    = 256 - lzThresh + lzF // 314
	tNode    = nChar*2 - 1          // 627
	rNode    = tNode - 1            // 626
	maxFreq  = 0x8000
)

var (
	dCode [256]int
	dLen  [256]int
)

func init() {
	// build d_code / d_len from StartHuff's count tables
	c210 := [6]int{1, 3, 8, 12, 24, 16}
	c216 := [6]int{32, 48, 64, 48, 48, 16}
	eax, bit, idx, ci := 0, 0x20, 0, 0
	for bit != 0 {
		ch := c210[ci]
		ci++
		for ch > 0 {
			for k := 0; k < bit; k++ {
				dCode[idx] = eax >> 22
				idx++
			}
			eax += 0x400000
			ch--
		}
		bit >>= 1
	}
	idx = 0
	for v, cnt := range c216 {
		for k := 0; k < cnt; k++ {
			dLen[idx] = v + 1
			idx++
		}
	}
}

type huff struct {
	freq [tNode + 1]int
	son  [tNode]int
	prnt [tNode + nChar]int
}

func newHuff() *huff {
	h := &huff{}
	for i := 0; i < nChar; i++ {
		h.freq[i] = 1
		h.son[i] = i + tNode
		h.prnt[i+tNode] = i
	}
	i, j := 0, nChar
	for j <= rNode {
		h.freq[j] = h.freq[i] + h.freq[i+1]
		h.son[j] = i
		h.prnt[i] = j
		h.prnt[i+1] = j
		i += 2
		j++
	}
	h.freq[tNode] = 0xFFFF
	h.prnt[rNode] = 0
	return h
}

func (h *huff) reconst() {
	j := 0
	for i := 0; i < tNode; i++ {
		if h.son[i] >= tNode {
			h.freq[j] = (h.freq[i] + 1) / 2
			h.son[j] = h.son[i]
			j++
		}
	}
	i := 0
	j = nChar
	for j < tNode {
		f := h.freq[i] + h.freq[i+1]
		h.freq[j] = f
		k := j - 1
		for f < h.freq[k] {
			k--
		}
		k++
		for m := j; m > k; m-- {
			h.freq[m] = h.freq[m-1]
			h.son[m] = h.son[m-1]
		}
		h.freq[k] = f
		h.son[k] = i
		i += 2
		j++
	}
	for i := 0; i < tNode; i++ {
		k := h.son[i]
		if k >= tNode {
			h.prnt[k] = i
		} else {
			h.prnt[k] = i
			h.prnt[k+1] = i
		}
	}
}

func (h *huff) update(c int) {
	if h.freq[rNode] == maxFreq {
		h.reconst()
	}
	c = h.prnt[c+tNode]
	for {
		h.freq[c]++
		k := h.freq[c]
		l := c + 1
		if k > h.freq[l] {
			for k > h.freq[l+1] {
				l++
			}
			h.freq[c] = h.freq[l]
			h.freq[l] = k
			i := h.son[c]
			h.prnt[i] = l
			if i < tNode {
				h.prnt[i+1] = l
			}
			j := h.son[l]
			h.son[l] = i
			h.prnt[j] = c
			if j < tNode {
				h.prnt[j+1] = c
			}
			h.son[c] = j
			c = l
		}
		c = h.prnt[c]
		if c == 0 {
			break
		}
	}
}

// lzhufDecompress decodes an LZHUF stream to exactly outSize bytes.
func lzhufDecompress(src []byte, outSize int) []byte {
	h := newHuff()
	ring := make([]byte, lzN)
	for i := range ring {
		ring[i] = 0x20
	}
	ringpos := 0xFC4
	out := make([]byte, 0, outSize)
	n := len(src)
	si := 0
	var ebp uint32
	cl := 8

	rd := func() uint32 {
		var b uint32
		if si < n {
			b = uint32(src[si])
		}
		si++
		return b
	}

	for len(out) < outSize {
		c := h.son[rNode]
		cl--
		for c < tNode {
			cl++
			if cl >= 0 {
				for {
					ebp |= rd() << uint(cl)
					cl -= 8
					if cl < 0 {
						break
					}
				}
			}
			bit := int((ebp >> 15) & 1)
			ebp = (ebp & 0xFFFF0000) | ((ebp << 1) & 0xFFFF)
			c = h.son[c+bit]
		}
		cl++
		c -= tNode
		h.update(c)

		if c < 256 {
			out = append(out, byte(c))
			ring[ringpos] = byte(c)
			ringpos = (ringpos + 1) & (lzN - 1)
			continue
		}
		// match: decode position
		if cl >= 0 {
			for {
				ebp |= (rd() & 0xFF) << uint(cl)
				cl -= 8
				if cl < 0 {
					break
				}
			}
		}
		ebp <<= 8
		i := int((ebp >> 16) & 0xFF)
		ch := dLen[i]
		base := dCode[i]
		cl += 8
		stateRefill := cl >= 0
		for {
			if stateRefill {
				for cl >= 0 {
					ebp |= (rd() & 0xFF) << uint(cl)
					cl -= 8
				}
			}
			ebp <<= 1
			ch--
			if ch == 0 {
				break
			}
			cl++
			stateRefill = cl >= 0
		}
		cl++
		pos := int((((ebp & 0x3F0000) | (uint32(base) << 22)) >> 16) & 0xFFFF)
		length := c - 253
		srcidx := (ringpos - 1 - pos) & (lzN - 1)
		for k := 0; k < length; k++ {
			b := ring[srcidx&(lzN-1)]
			srcidx++
			out = append(out, b)
			ring[ringpos] = b
			ringpos = (ringpos + 1) & (lzN - 1)
			if len(out) >= outSize {
				break
			}
		}
	}
	return out
}

File: remake/engine/lzhuf.py
"""
lzhuf.py — NGI LZHUF (method 0x80) decompressor.

Register-faithful port of NGI32.DLL sub_26F4E (bl & 0x80 branch) and its helpers
sub_26D00 (StartHuff), sub_26DCE (DecodeChar/update). This is LZSS + adaptive
Huffman, but with NGI's OWN position tables (d_code/d_len built by StartHuff),
which are NOT the textbook lzhuf tables — that difference is why a naive port
diverges on long streams.

Confirmed constants: N=4096, F=60, THRESHOLD=2, T=627, R=626, ring start=0xFC4,
match length = symbol-253.  Shared 32-bit bit accumulator: tree navigation
consumes bit 15 (`shl bp,1`); position decode reads a byte from bits 16-23
(`shl ebp,8`).  d_code/d_len below were emulated from StartHuff's build loops.
"""
from __future__ import annotations

N, F, THRESHOLD = 4096, 60, 2
N_CHAR = 256 - THRESHOLD + F      # 314
T = N_CHAR * 2 - 1                # 627
R = T - 1                         # 626
MAX_FREQ = 0x8000
MASK = 0xFFFFFFFF

# NGI position tables (from StartHuff build, sub_26D00). d_code values 0..63,
# d_len bit-lengths 1..6.  Built from count tables [1,3,8,12,24,16]/[32,48,64,48,48,16].
_C210 = (1, 3, 8, 12, 24, 16)
_C216 = (32, 48, 64, 48, 48, 16)


def _build_tables():
    d_code = [0] * 256
    eax = 0
    bit = 0x20
    idx = 0
    ci = 0
    while bit:
        ch = _C210[ci]; ci += 1
        while ch > 0:
            for _ in range(bit):
                d_code[idx] = eax >> 22
                idx += 1
            eax += 0x400000
            ch -= 1
        bit >>= 1
    d_len = []
    val = 1
    for cnt in _C216:
        d_len += [val] * cnt
        val += 1
    return d_code, d_len


D_CODE, D_LEN = _build_tables()


class _Huff:
    """Adaptive Huffman tree — StartHuff/update/reconst (sub_26D00 / sub_26DCE)."""

    def __init__(self):
        self.freq = [0] * (T + 1)
        self.son = [0] * T
        self.prnt = [0] * (T + N_CHAR)
        for i in range(N_CHAR):
            self.freq[i] = 1
            self.son[i] = i + T
            self.prnt[i + T] = i
        i, j = 0, N_CHAR
        while j <= R:
            self.freq[j] = self.freq[i] + self.freq[i + 1]
            self.son[j] = i
            self.prnt[i] = self.prnt[i + 1] = j
            i += 2
            j += 1
        self.freq[T] = 0xFFFF
        self.prnt[R] = 0

    def reconst(self):
        freq, son, prnt = self.freq, self.son, self.prnt
        j = 0
        for i in range(T):
            if son[i] >= T:
                freq[j] = (freq[i] + 1) // 2
                son[j] = son[i]
                j += 1
        i = 0
        j = N_CHAR
        while j < T:
            f = freq[j] = freq[i] + freq[i + 1]
            k = j - 1
            while f < freq[k]:
                k -= 1
            k += 1
            for m in range(j, k, -1):
                freq[m] = freq[m - 1]
                son[m] = son[m - 1]
            freq[k] = f
            son[k] = i
            i += 2
            j += 1
        for i in range(T):
            k = son[i]
            if k >= T:
                prnt[k] = i
            else:
                prnt[k] = prnt[k + 1] = i

    def update(self, c):
        freq, son, prnt = self.freq, self.son, self.prnt
        if freq[R] == MAX_FREQ:
            self.reconst()
        c = prnt[c + T]
        while True:
            freq[c] += 1
            k = freq[c]
            l = c + 1
            if k > freq[l]:
                while k > freq[l + 1]:
                    l += 1
                freq[c] = freq[l]
                freq[l] = k
                i = son[c]
                prnt[i] = l
                if i < T:
                    prnt[i + 1] = l
                j2 = son[l]
                son[l] = i
                prnt[j2] = c
                if j2 < T:
                    prnt[j2 + 1] = c
                son[c] = j2
                c = l
            c = prnt[c]
            if c == 0:
                break


def decompress(src: bytes, out_size: int) -> bytes:
    huff = _Huff()
    son = huff.son
    ring = bytearray([0x20] * N)
    ringpos = 0xFC4
    out = bytearray()
    n = len(src)
    si = 0
    ebp = 0        # 32-bit bit accumulator
    cl = 8

    while len(out) < out_size:
        # --- DecodeChar: navigate tree (entry 0x26FAD) ---
        c = son[R]
        cl -= 1
        while c < T:
            cl += 1
            if cl >= 0:                       # refill (0x26FB8)
                while True:
                    b = src[si] if si < n else 0
                    si += 1
                    ebp = (ebp | (b << cl)) & MASK
                    cl -= 8
                    if cl < 0:
                        break
            # getbit from bit 15 (0x26FCA: shl bp,1)
            bit = (ebp >> 15) & 1
            ebp = (ebp & 0xFFFF0000) | ((ebp << 1) & 0xFFFF)
            c = son[c + bit]
        cl += 1                               # 0x26FE6
        c -= T
        huff.update(c)

        if c < 256:                           # literal
            out.append(c)
            ring[ringpos] = c
            ringpos = (ringpos + 1) & (N - 1)
            continue

        # --- match: decode position (0x27022) ---
        if cl >= 0:                           # top-up (0x2702B)
            while True:
                b = src[si] if si < n else 0
                si += 1
                ebp = (ebp | ((b & 0xFF) << cl)) & MASK
                cl -= 8
                if cl < 0:
                    break
        ebp = (ebp << 8) & MASK               # 0x27040
        i = (ebp >> 16) & 0xFF
        ch = D_LEN[i]
        base = D_CODE[i]
        cl += 8                               # 0x2705A
        state_refill = (cl >= 0)
        while True:
            if state_refill:                  # L_refill (0x27069)
                while cl >= 0:
                    b = src[si] if si < n else 0
                    si += 1
                    ebp = (ebp | ((b & 0xFF) << cl)) & MASK
                    cl -= 8
                state_refill = False
            ebp = (ebp << 1) & MASK           # L_shift (0x2705F)
            ch -= 1
            if ch == 0:
                break
            cl += 1
            state_refill = cl >= 0
        cl += 1                               # 0x2707E inc cl
        pos = (((ebp & 0x3F0000) | (base << 22)) >> 16) & 0xFFFF

        length = c - 253                      # symbol-253  (0x2709A lea ebp,[ebx-0xFD])
        srcidx = (ringpos - 1 - pos) & (N - 1)
        for _ in range(length):
            b = ring[srcidx & (N - 1)]
            srcidx += 1
            out.append(b)
            ring[ringpos] = b
            ringpos = (ringpos + 1) & (N - 1)
            if len(out) >= out_size:
                break

    return bytes(out)
